sql/2020/03_Markup/pages_markup_by_device_and_button_types.sql

```diff
@@ -0,0 +1,44 @@
+#standardSQL
+# pages markup metrics grouped by device and button type
+
+# helper to create percent fields
+CREATE TEMP FUNCTION AS_PERCENT (freq FLOAT64, total FLOAT64) RETURNS FLOAT64 AS (
+  ROUND(SAFE_DIVIDE(freq, total), 4)
+);
+
+# returns all the data we need from _markup
+CREATE TEMPORARY FUNCTION get_markup_buttons_info(markup_string STRING)
+RETURNS ARRAY<STRUCT<
+name STRING,
+freq INT64
+>> LANGUAGE js AS '''
+var result = [];
+try {
+    var markup = JSON.parse(markup_string); 
+
+    if (Array.isArray(markup) || typeof markup != 'object') return result;
+
+    if (markup.buttons && markup.buttons.types) {
+      return Object.entries(markup.buttons.types).map(([name, freq]) => ({name, freq})); 
+    }
+
+} catch (e) {}
+return result;
+''';
+
+SELECT
+  _TABLE_SUFFIX AS client,
+  button_type_info.name AS button_type,
+  SUM(button_type_info.freq) AS freq, 
```

Comment:
Do you intend to make the frequency the number of occurrences of the button type, or the number of pages having that button type? I think the latter may be more meaningful to readers since it doesn't vary by pages that may use a button type many many times.

Reply:
I think I've worked this one out. I struggle with these totals and frequencies:

```
  _TABLE_SUFFIX AS client,
  #total AS pages_total,
  button_type_info.name AS button_type,
  COUNTIF(button_type_info.freq > 0) AS freq_page_with_button, 
  AS_PERCENT(COUNTIF(button_type_info.freq > 0), total) AS pct_page_with_button,
  SUM(button_type_info.freq) AS freq_button, 
  #SUM(SUM(button_type_info.freq)) OVER (PARTITION BY _TABLE_SUFFIX) as buttons_total,
  AS_PERCENT(SUM(button_type_info.freq), SUM(SUM(button_type_info.freq)) OVER (PARTITION BY _TABLE_SUFFIX)) AS pct_button
```

pct_page_with_button should be based on pages that contain the button at least once. What you suggested.
pct_button is using the raw count of buttons, so is more like a ratio for the types of buttons used.

Would that make sense?

Reply:
I've submitted the change (minus commented out lines)